### ISSUE TITLE
[SPIKE needed] Ergonomic Improvement Ideas

### ISSUE THREAD

**Issue by LevelbossMike** [LevelbossMike/ember-statecharts#235]
As discussed with @nullvoxpopuli, @pangratz and @sukima in a recent call we want to try to improve the ergonomics of the addon for modern Ember.js usage.

Areas for improvement:

* make it __easy__ and __intuitive__ to use __all__ features from xstate in an Ember.js application
* allow passing machines directly and not necessarily create an interpreter via the `statechart`-decorator.
* don't hide xstate behind custom addon related "magic"

We came up with some ideas about a new way to integrate xstate with `ember-statecharts`:

## Using a mechanism similiar to [`@use and Resources`- RFC](https://github.com/emberjs/rfcs/pull/567)

```typescript
import Component from '@glimmer/component';
import { useMachine, matchesState } from 'ember-statecharts';
// to extract machine configs to a different place than actual component code
import TestMachine from '../machines/test';


export default class Test extends Component {
  /*
   @use does not exist yet. We could imagine building something similar
   ourselves for now to emulate its behavior for now - maybe we could call
   it @useMachine
  */
  @use statechart = useMachine(Test)
    .withConfig({ /* ... */}) // to allow calling component to customize `Test`-machine
    .withContext({/*...*/}); // to allow calling component to customize `Test`-machine

  // still need tbd if we want to keep @matchesState
  @matchesState('testing')
  isTesting: boolean;
}
```

**Benefits of this approach**

* aligned with the future vision of the framework, i.e. `@use`
* we can hook into the component lifecycle to start the underlying xstate-interpreter and clean it up when the component gets unrendered / destroyed
* it looks very similar to what other framework wrappers of xstate do - e.g. [xstate/react](https://github.com/davidkpiano/xstate/tree/master/packages/xstate-react) or [xstate/vue](https://github.com/davidkpiano/xstate/tree/master/packages/xstate-vue)

**Questions**
* how can we emulate the `@use`-decorator?
* does it make sense to write an `@use`-like functionality just for this addon that we can later on switch out for `@use`?

**Next steps**
* Do a spike implementation to see if it is possible to implement this api with the current primitives the language and framework provide.

@nullvoxpopuli, @pangratz, @sukima as far as I remember we were all pretty much in favor of this approach so I'm adding the other versions we discussed only for reference.

#### class decorator

```typescript
import Component from '@glimmer/component';
import TestMachine from '../machines/test';

@useMachine(TestMachine)
export default class Test extends Component {
  @withConfig
  get() {
    return {
      actions: {
        handleSuccess: this.handleSuccess
      }
    }
  }

  @withContext
  get() {
    return {
      count: 1
    }
  }

  @statechartAction
  handleSuccess(context: any, event: { type: string, ...}) {
    // context.count => 1
  }

  @transition
  inputChanged(send: () => {}, event: { target: { value }}) {
    send('INPUT_CHANGED', { value });
  }
}
```

#### Property decorator

```typescript
import Component from '@glimmer/component';
import TestMachine from '../machines/test';

export default class Test extends Component {
  @useMachine(TestMachine)
  interpreter;

  @withConfig
  get() {
    return {
      actions: {
        handleSuccess: this.handleSuccess
      }
    }
  }

  @withContext
  get() {
    return {
      count: 1
    }
  }

  @statechartAction
  handleSuccess(context: any, event: { type: string, ...}) {
    // context.count => 1
  }

  @action
  inputChanged(event: { target: { value } }) {
    this.interpreter.send('INPUT', { value });
  }
}
```

**Comment by NullVoxPopuli** [comment]
In addition, we should explore different ways of handling ember-concurrency tasks as machine-invokables.

@sukima has a wonderful little utility here that enables tasks to be located anywhere outside the lifecycle of the machine: https://tritarget.org/#ember-concurrency%20with%20XState:[[ember-concurrency%20with%20XState]]%20[[Ember%20XState%20Utility]]
(which is handy for my use case, where I'm defining tasks on ephemeral classes that aren't exactly tied to component-state (though that's easier now with the destroyables polyfill / RFC).

It may be beneficial to support the scenario where a nested component contains a task that the machine is aware of and _something_ cancels the task. Supporting both cancellation and error would allow for slightly different workflows, if needed.

**Comment by LevelbossMike** [comment]
Current state:

https://github.com/LevelbossMike/ember-statecharts/blob/284c8c92894c7fa4837fc7d6ffc8d57816008aac/tests/unit/use-machine-test.js

It works and makes use of https://github.com/pzuraq/ember-usable under the hood

**Comment by NullVoxPopuli** [comment]
cc @pzuraq 
VERY EXCITE

**Comment by pzuraq** [comment]
hmm, I need to finish that, [this branch](https://github.com/pzuraq/ember-usable/tree/initial-implementation) is where I'm taking it but haven't had time to finish it up and land it.

Seems like a perfect use case for `@use` though 😄

**Comment by alexlafroscia** [comment]
I’m curious about the syntax like

```javascript
  @matchesState(‘testing’)
  IsTesting: Boolean
```

Reading something like that throws me off; is reminds me of something like that the `eq` decorator and friends, but there’s no explicit pointer do the state machine that’s in the `testing` state. I guess it assumes that the chart itself is available as the `statechart` property, but does it make sense for the API to make that assumption? To me at least, it makes things harder to read as I end up asking myself “_what_ matches this state?”

**Comment by NullVoxPopuli** [comment]
> but there’s no explicit pointer do the state machine

yeah, that's my feeling about it too. And then what happens if someone wants 2 or more state machines in a component? those decorators wouldn't be able to work.

I think exporting some vanilla JS helpers will help a bunch, as well as some helpers to use in the template. I think there should be some explicitness about which machine is being used where. idk.
I'm personally _uncomfortable_ with forcing a single machine to one component

**Comment by alexlafroscia** [comment]
> happens if someone wants 2 or more state machines in a component? those decorators wouldn't be able to work.

That’s a second concern of mine as well that goes right along with my confusion. I’ll admit that I know nothing about `xstate` so maybe “one component manipulating two state charts” is just bad design, but I have a hunch that’s not something that should be totally off-the-books. If a redesign of the API for the library is in discussion, I’d love to see this concern explored! Even just expanding the API to something more like

```javascript
@matchesState(‘statechart’, ‘testing’)
IsTesting
```

Would be much easier to grok on a quick read without context (which will be a lot of people reading most code!)

**Comment by LevelbossMike** [comment]
> I’m curious about the syntax like
> 
> ```js
>   @matchesState(‘testing’)
>   IsTesting: Boolean
> ```
> 
> Reading something like that throws me off; it reminds me of something like that the `eq` decorator and friends, but there’s no explicit pointer to the state machine that’s in the `testing` state. I guess it assumes that the chart itself is available as the `statechart` property, but does it make sense for the API to make that assumption? To me at least, it makes things harder to read as I end up asking myself “_what_ matches this state?”

The idea is that you usually only have one so it's easier to just have a default property that it matches against. Right now the default is `statechart`. But it already works for a different property name even if you had two statecharts in your component:

https://github.com/LevelbossMike/ember-statecharts/blob/master/tests/unit/computed-test.js#L155

`matchesState('started', 'testing')`

Will match against the `testing`-xstateService.

It might make sense to have configuration option for this though. So if you call you statechart properties different you might want to change the default in one place instead of always doing `@matchesState(<state>, 'myCustomStatechartPropertyNameIUseEverywhere')`.

If you don't like this implicitness you can always do `@matchesStates(<state>, 'statechart')` already.

**Comment by alexlafroscia** [comment]
> If you don't like this implicitness you can always do @matchesStates(<state>, 'statechart') already.

That’s awesome!

>  It might makes sense to change the docs to the explicit version - because many people share this feeling when getting into ember-statecharts for the first time. 🤔

That makes a lot of sense to me! Sorry to derail the conversation from the initial purpose, but I thought that addressing that feedback might be tied up with addressing the API of the addon as a whole.

**Comment by sukima** [comment]
The idea of these decorators is to wrap the boilerplate up in a neat tidy package. Thus such I would argue one component running two or more machines is out of scope and to implement that yourself. The boilerplate to interface with XState is so minimal it is very easy to use it out-of-the-box. That said one could also just have two `EmberObject`s each with their own `@useMachine` and one component that manages them.

There are so many ways to skin the cat there does not need to have [one twue way](https://thoughtcatalog.files.wordpress.com/2015/04/zciblll.gif?w=500&h=210).
